paddle/platform/enforce.h

```diff
@@ -0,0 +1,160 @@
+ * __LINE__
+ *
+ * This macro take __VA_ARGS__, user can pass any type if that type can
+ * serialize to std::ostream
+ */
+#define PADDLE_THROW(...)                                            \
+  do {                                                               \
+    throw ::paddle::platform::EnforceNotMet(                         \
+        ::paddle::string::Sprintf(__VA_ARGS__), __FILE__, __LINE__); \
+  } while (0)
+
+#ifndef PADDLE_ONLY_CPU
+
+template <typename... Args>
+inline void throw_on_error(cudaError_t e, const Args&... args) {
+  if (e) {
+    std::stringstream ss;
+    ss << ::paddle::string::Sprintf(args...);
+    ss << ::paddle::string::Sprintf(" at [%s:%s];", __FILE__, __LINE__);
+    throw thrust::system_error(e, thrust::cuda_category(), ss.str());
+  }
+}
+
+template <typename... Args>
+inline void throw_on_error(curandStatus_t stat, const Args&... args) {
+  if (stat != CURAND_STATUS_SUCCESS) {
+    std::stringstream ss;
+    ss << ::paddle::string::Sprintf(args...);
+    ss << ::paddle::string::Sprintf(" at [%s:%s];", __FILE__, __LINE__);
+    throw thrust::system_error(cudaErrorLaunchFailure, thrust::cuda_category(),
+                               ss.str());
+  }
+}
+
+template <typename... Args>
+inline void throw_on_error(cudnnStatus_t stat, const Args&... args) {
+  if (stat == CUDNN_STATUS_SUCCESS) {
+    return;
+  } else {
+    std::stringstream ss;
+    ss << ::paddle::platform::dynload::cudnnGetErrorString(stat);
+    ss << ", " << ::paddle::string::Sprintf(args...);
+    ss << ::paddle::string::Sprintf(" at [%s:%s];", __FILE__, __LINE__);
+    throw std::runtime_error(ss.str());
+  }
+}
+
+template <typename... Args>
+inline void throw_on_error(cublasStatus_t stat, const Args&... args) {
+  std::stringstream ss;
+  if (stat == CUBLAS_STATUS_SUCCESS) {
+    return;
+  } else if (stat == CUBLAS_STATUS_NOT_INITIALIZED) {
+    ss << "CUBLAS: not initialized";
+  } else if (stat == CUBLAS_STATUS_ALLOC_FAILED) {
+    ss << "CUBLAS: alloc failed";
+  } else if (stat == CUBLAS_STATUS_INVALID_VALUE) {
+    ss << "CUBLAS: invalid value";
+  } else if (stat == CUBLAS_STATUS_ARCH_MISMATCH) {
+    ss << "CUBLAS: arch mismatch";
+  } else if (stat == CUBLAS_STATUS_MAPPING_ERROR) {
+    ss << "CUBLAS: mapping error";
+  } else if (stat == CUBLAS_STATUS_EXECUTION_FAILED) {
+    ss << "CUBLAS: execution failed";
+  } else if (stat == CUBLAS_STATUS_INTERNAL_ERROR) {
+    ss << "CUBLAS: internal error";
+  } else if (stat == CUBLAS_STATUS_NOT_SUPPORTED) {
+    ss << "CUBLAS: not supported";
+  } else if (stat == CUBLAS_STATUS_LICENSE_ERROR) {
+    ss << "CUBLAS: license error";
+  }
+  ss << ", " << ::paddle::string::Sprintf(args...);
+  ss << ::paddle::string::Sprintf(" at [%s:%s];", __FILE__, __LINE__);
+  throw std::runtime_error(ss.str());
+}
+
+#endif  // PADDLE_ONLY_CPU
+
+template <typename... Args>
+inline void throw_on_error(int stat, const Args&... args) {
```

Comment:
I will double check this, because we already use `bool` condition in our tensor implementation, but it didn't throw a error.